vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,680 @@
+legend("left", lwd = 1, col = c("#999966", "#8c8cd9", "#cc0044"), 
+       legend = c("S", "I", "R"), bty = "n")
+```
+
+Typically you will not use the `run()` function of the particle filter directly, and instead use the functions below to infer parameters.
+
+### Using MCMC to infer parameters
+
+Using this, we can do a MCMC using the Metropolis-Hastings algorithm to infer the values of $\beta$ and $\gamma$ from daily case counts. Let's first describe the parameters we wish to infer, giving a minimum, maximum. For $\gamma$ we will also apply a gamma prior with shape 1 and rate 0.2: $\Gamma(1, 0.2)$:
+
+```{r}
+beta <- pmcmc_parameter("beta", 0.2, min = 0)
+gamma <- pmcmc_parameter("gamma", 0.1, min = 0, 
+                         prior = function(p) {
+                           dgamma(p, shape = 1, scale = 0.2, log = TRUE)
+                         })
+
+proposal_matrix <- matrix(c(0.01^2, 0, 0, 0.01^2), nrow = 2, ncol = 2, byrow = TRUE)
+mcmc_pars <- pmcmc_parameters$new(list(beta = beta, gamma = gamma), proposal_matrix)
+```
+
+The proposals for each MCMC step are also set here. Proposals are drawn from a multi-variate normal distribution, with the variance-covariance matrix set here. In this case we have $\beta_{n+1} \sim N(\beta_{n}, 0.01^2)$ and $\gamma_{n+1} \sim N(\gamma_{n}, 0.01^2)$ i.e. no covariance.
+
+Finally, the sampler can be run using these parameters, and the particle filter defined above. Taking 2000 steps with two independent chains:
+
+```{r pmcmc_combined, cache=TRUE}
+pmcmc_run <- 
+  pmcmc(
+    mcmc_pars,
+    filter,
+    2000,
+    save_state = TRUE,
+    save_trajectories = TRUE,
+    progress = TRUE,
+    n_chains = 2
+  )
+```
+
+We also provide some basic tools for dealing with this output, before running prediction (see below). We can remove burn-in and thin the four chains as follows:
+
+```{r}
+processed_chains <- pmcmc_thin(pmcmc_run, burnin = 500, thin = 2)
+processed_chains
+apply(processed_chains$pars, 2, mean)
+```
+
+### Optimal use of parallelisation
+
+Note that we run four separate chains in serial here. If you have $m$ nodes available, each with $n$ cores, the most efficient way to run this step is to set `nthreads = n` on the filters, and run $m$ independent chains, one on each node setting `n_chains = 1`. These independent results can then be collected using `pmcmc_combine()`:
+
+```{r pmcmc_serial, cache=TRUE}
+pmcmc1 <- 
+  pmcmc(
+    mcmc_pars,
+    filter,
+    2000,
+    save_state = TRUE,
+    save_trajectories = FALSE,
+    progress = TRUE,
+    n_chains = 1
+  )
+pmcmc2 <- 
+  pmcmc(
+    mcmc_pars,
+    filter,
+    2000,
+    save_state = TRUE,
+    save_trajectories = FALSE,
+    progress = TRUE,
+    n_chains = 1
+  )
+pmcmc3 <- 
+  pmcmc(
+    mcmc_pars,
+    filter,
+    2000,
+    save_state = TRUE,
+    save_trajectories = FALSE,
+    progress = TRUE,
+    n_chains = 1
+  )
+pmcmc_distributed_run <- pmcmc_combine(pmcmc1, pmcmc2, pmcmc3)
```

Comment:
Why is this called "distributed" if it was run in series?

Reply:
ah, ok the idea is that these might have been run across different computers? This feels too much a detail of our workflow for an introductory vignette and might fit better in a vignette on scaling. It also avoids the very real problem of correctly getting the RNG streams correct here.

Reply:
With all the above, I'd suggest dropping this section for now

Reply:
Moved to another vignette, we can decide whether to keep or remove separately. I think it's an important advantage, so worth having somewhere from the get go. Agree it's too involved for the intro though